### DISCUSSION TITLE
Type error when instantiating class from __subclasses__() of an ABC parent

### DISCUSSION THREAD

**Discussion by argilya** [microsoft/pyright#5747] [Q&A] [answered]
I am curious how I should approach this typing error:

```python
# flake8: noqa
from abc import ABC, abstractmethod


class Example(ABC):
    @abstractmethod
    def method(self) -> None:
        raise NotImplementedError


class Child(Example):
    def method(self) -> None:
        print("I'm a child!")


for subcls in Example.__subclasses__():
    instance = subcls()
    
    # Cannot instantiate abstract class "Example"
    # "Example.method" is abstract (Pyright reportGeneralTypeIssues)
```

Pyright thinks that `subcls` is `Type[Example]`, and `Example` can't be instantiated because it is abstract. I would have expected Pyright to recognize that `__subclass__()` returns concrete classes.

What am I doing wrong? How do I resolve this error? Thank you!

**Comment by erictraut** [comment, marked as answer]
Pyright has no special knowledge of `__subclass__`. It gets its information from type stubs. Here's the relevant portion of the type declaration for the `type` class:

```python
class type:
    ...
    def __subclasses__(self: _typeshed.Self) -> list[_typeshed.Self]: ...
```

Not surprisingly, mypy gives the same error in this case.

Even if this were special-cased, I don't think it's a safe assumption that it always "returns concrete classes". Subclasses of ABCs can still be abstract.

I recommend using a `# type: ignore` or `# pyright: ignore` comment to suppress the error in this case — or to avoid using the `__subclasses__` call altogether.

**Reply by argilya** [reply]
> Subclasses of ABCs can still be abstract.

I failed to realize this. Thank you!
